File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/passwordmanager/passwordinfo/PasswordInfo.java
package org.elastos.trinity.runtime.passwordmanager.passwordinfo;

import org.elastos.trinity.runtime.passwordmanager.PasswordType;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Root type for all password information. This type is abstract and should not be used
 * directly.
 */
public abstract class PasswordInfo {
    /**
     * Unique key, used to identity the password info among other.
     */
    public String key;

    /**
     * Password type, that defines the format of contained information.
     */
    public PasswordType type;

    /**
     * Name used while displaying this info. Either set by users in the password manager app
     * or by apps, when saving passwords automatically.
     */
    public String displayName;

    /**
     * List of any kind of app-specific additional information for this password entry.
     */
    public JSONObject custom;

    public JSONObject asJsonObject() {
        try {
            JSONObject jsonObj = new JSONObject();
            jsonObj.put("key", key);
            jsonObj.put("type", type.mValue);
            jsonObj.put("displayName", displayName);
            jsonObj.put("custom", custom);
            return jsonObj;
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    public void fillWithJsonObject(JSONObject jsonObj) throws Exception {
        if (!jsonObj.has("key") || !jsonObj.has("type") || !jsonObj.has("displayName"))
            throw new Exception("Invalid password info, some base fields are missing");

        this.key = jsonObj.getString("key");
        this.type = PasswordType.fromValue(jsonObj.getInt("type"));
        this.displayName = jsonObj.getString("displayName");

        if (jsonObj.has("custom")) {
            this.custom = jsonObj.getJSONObject("custom");
        }
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/PasswordManager/PasswordInfo/PasswordInfo.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import Foundation

/**
 * Root type for all password information. This type is abstract and should not be used
 * directly.
 */
public class PasswordInfo {
    /**
     * Unique key, used to identity the password info among other.
     */
    public var key: String = ""

    /**
     * Password type, that defines the format of contained information.
     */
    public var type: PasswordType = .GENERIC_PASSWORD

    /**
     * Name used while displaying this info. Either set by users in the password manager app
     * or by apps, when saving passwords automatically.
     */
    public var displayName: String = ""

    /**
     * Package ID of the application/capsule that saved this password information.
     * READ-ONLY
     */
    public var appID: String? = nil

    /**
     * List of any kind of app-specific additional information for this password entry.
     */
    public var custom: Dictionary<String, Any>? = nil

    public func asDictionary() -> Dictionary<String, Any>? {
        var jsonObj = Dictionary<String, Any>()
        jsonObj["key"] = key
        jsonObj["type"] = type
        jsonObj["displayName"] = displayName
        jsonObj["custom"] = custom
        jsonObj["appID"] = appID
        return jsonObj
    }

    public func fillWithJsonObject(_ jsonObj: Dictionary<String, Any>) throws {
        if !jsonObj.keys.contains("key") || !jsonObj.keys.contains("type") || !jsonObj.keys.contains("displayName") {
            throw "Invalid password info, some base fields are missing"
        }
        
        self.key = jsonObj["key"] as! String
        self.type = PasswordType.init(rawValue: jsonObj["type"] as! Int)!
        self.displayName = jsonObj["displayName"] as! String

        if jsonObj.keys.contains("custom") {
            self.custom = jsonObj["custom"] as? Dictionary<String, Any>
        }
       
        // SECURITY NOTE - Don't fill with appID. AppID is filled automatically.
    }
}
